fix: added connection error handling

sf3ris opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

#### Description
As discussed in issue #95 we should check the connection error when registering the plugin.


#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

mcollina commented:
Good work! Could you update the docs?

mcollina reviewed (approved):
lgtm

zekth commented on README.md:
```diff
@@ -186,6 +186,86 @@ fastify.listen(3000, function (err) {
+
+const fastify = require('fastify')()
+
+// lazyConnect option will not estabilish the connection until the first command is issued
```
```suggestion
// lazyConnect option will not establish the connection until the first command is issued
```

zekth reviewed (changes requested):
Few typos otherwise LGTM

climba03003 commented on README.md:
````diff
@@ -186,6 +186,86 @@ fastify.listen(3000, function (err) {
+})
+```
+```
+Error: ERR_AVVIO_PLUGIN_TIMEOUT: plugin did not start in time: /fastify-redis/index.js. You may have forgotten to call 'done' function or to resolve a Promise
````
Do we need to show the full error message inside the README? I think a sentence at start is enough.

sf3ris replied:
Dont know tbh. :thinking: 
I thought could be useful to differentiate the two different errors that the user could encounter and explaining why.
If you have a strong opinion regarding this we can for sure keep only the first sentence.

climba03003 replied:
I do not have strong opinion on it. Just wondering what make you think the error should exist inside the README. If you think it will be useful, I am totally fine with it.

sf3ris replied:
I think you were right about this, no need to show error examples on README. Thanks for the tip :pray: See commit [d3064c0](https://github.com/fastify/fastify-redis/pull/98/commits/d3064c00b2bdcb916d75ba0301066983a2c3a7b3)

zekth reviewed (approved):
lgtm nice

Eomm commented on index.js:
```diff
@@ -64,7 +64,13 @@ function fastifyRedis (fastify, options, next) {
-  next()
+  if (!redisOptions.lazyConnect) {
+    return client.info((err, _) => {
+      return err ? next(new Error('fastify-redis tried to connect to an invalid redis instance')) : next()
```
I would not loose the origin `err` here.
We could set the error cause

sf3ris replied:
I wanted to hide the original `err` because in the majority of the cases the error will be 
```
MaxRetriesPerRequestError: Reached the max retries per request limit (which is 1). Refer to "maxRetriesPerRequest" option for details.
```
I dont know if it is clear enough when the real error is that the connection cannot be established. What you think about it?

Eomm replied:
What would be the error in case of the wrong password?
Moreover if the user google that error, the user will find ioredis related issues that is good I think

sf3ris replied:
Yeah i think you're right. See commit [a467f07](https://github.com/fastify/fastify-redis/pull/98/commits/a467f07ccc09393f0db473c3213f17af85ee05ba)

sf3ris commented:
I've added the error propagation from the ioredis library when trying to connect and added tests to simulate the behavior when connecting on an auth protected redis instance. See commit [a467f07](https://github.com/fastify/fastify-redis/pull/98/commits/a467f07ccc09393f0db473c3213f17af85ee05ba).

climba03003 commented on README.md:
````diff
@@ -186,6 +186,86 @@ fastify.listen(3000, function (err) {
+})
+```
+```
+Error: fastify-redis tried to connect to an invalid redis instance
````
I think it should be updated in favor of new update.

climba03003 reviewed (changes requested):
The CI test is failing.

mcollina reviewed (approved):
lgtm

climba03003 commented on .github/workflows/ci.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
     services:
       redis:
-        image: redis
+        image: redis:5
```
why it is not the latest? if validating the error message is not possible or hard to maintain. change the test assertion to `t.ok(err)` is fine (it means error must exist).

sf3ris replied:
I kinda disagree this time, I think that relying on the `latest` tag of an image is not so good, because in the next major they can totally change their APIs and i dont expect the whole CI to fail due to this, i would update the version incrementally when needed and test the behavior with the new major releases. 
Futhermore i would expect to work with the same version between the local docker environment and the CI one.

I had bad episodes with `latest` tag, it was really hard to debug when something was silently failing. Does this make sense to you? Or still prefering the `latest` tag?

climba03003 replied:
The document stated that we are required for `redis@5` or greater and the CI is failing for `redis@latest`. I am fine for a fixed version, but I think the test should pass for both `redis@5` and `redis@6` currently.

sf3ris replied:
Thats a really nice point :+1: How would u tackle this? testing it locally with different versions or implementing inside the CI?
Im fine to revert back to lastest tho, but i had so many ghosts with it :smiley: 

BTW I agree on generalizing the error inside the tests, uploading a patch soon. Thanks for the tips.

climba03003 replied:
I think we can make use of the `strategy.matrix`, technically it may pass to `services` like `steps` below. But I haven't tested it.

e.g.
```
strategy:
      matrix:
        node-version: [10, 12, 14, 16]
        redis-version: [5, 6]
    services:
      redis:
        image: redis:${{ matrix.redis-version }}
```

sf3ris replied:
Yeah came up with the same idea, testing a CI to see if it feasible.

climba03003 commented on .github/workflows/ci.yml:
```diff
@@ -17,9 +17,10 @@ jobs:
     strategy:
       matrix:
         node-version: [10, 12, 14, 16]
+        redis-tag: [5,6]
```
Space after comma for consistency.
```suggestion
        redis-tag: [5, 6]
```

climba03003 reviewed (approved):
LGTM. 
@Eomm do you have any more comments on it? If no, I will be merging it and release a new version.

Eomm reviewed (approved):
👍🏽
